File: app/src/main/java/com/mudrahome/MHLMS/Fragments/LeadDetailsFragment.java
package com.mudrahome.MHLMS.Fragments;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.NonNull;

import com.mudrahome.MHLMS.Firebase.Firestore;
import com.mudrahome.MHLMS.Interfaces.FirestoreInterfaces;
import com.mudrahome.MHLMS.Managers.PermissionManager;
import com.mudrahome.MHLMS.Models.TimeModel;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import android.text.Html;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.mudrahome.MHLMS.Managers.TimeManager;
import com.mudrahome.MHLMS.Models.LeadDetails;
import com.mudrahome.MHLMS.Models.UserDetails;
import com.mudrahome.MHLMS.Models.UserList;
import com.mudrahome.MHLMS.R;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

@SuppressLint("ValidFragment")
public class LeadDetailsFragment extends BottomSheetDialogFragment {

    private BottomSheetBehavior mBehavior;
    private TextView name;
    private TextView loan;
    private TextView number;
    private TextView employment;
    private TextView employmentType;
    private TextView loanType;
    private TextView propertyType;
    private TextView location;
    private TextView assignedTo;
    private TextView assigner;
    private TextView status;
    private TextView assignedOn;
    private TextView assignedAt;
    private TextView customerRemarks;
    private TextView workDate;
    private TextView workTime;
    private TextView bankNames;
    private TextView assignerContact;
    private TextView assigneeContact;

    private Button button;
    private LinearLayout assignedToLayout, assignerLayout, callerRemarksLayout, sallerRemarksLayout, employmentTypeLayout;
    private LinearLayout salesmanRemarksHeadingLayout;
    private LinearLayout customerRemarksLayout;
    private ProgressDialog progress;
    private Context context;

    private LeadDetails leadDetails;
    private com.mudrahome.MHLMS.Firebase.Firestore firestore;
    private BroadcastReceiver br;
    private String userType;

    private String customerNotInterested = "Customer Not Interested";
    private String documentPicked = "Document Picked";
    private String customerFollowUp = "Customer follow Up";
    private String customerNotContactable = "Customer Not Contactable";
    private String customerInterestedButDocumentPending = "Customer Interested but Document Pending";
    private String notDoable = "Not Doable";
    private String documentPickedFileLoggedIn = "Document Picked and File Logged in";




    public LeadDetailsFragment(LeadDetails leadDetails, Context context, String userType) {
        this.leadDetails = leadDetails;
        this.context = context;
        this.userType = userType;
    }

    @Override
    public void setupDialog(Dialog dialog, int style) {
        BottomSheetDialog bottomSheetDialog = (BottomSheetDialog) dialog;
        try {
            Field behaviorField = bottomSheetDialog.getClass().getDeclaredField("behavior");
            behaviorField.setAccessible(true);
            final BottomSheetBehavior behavior = (BottomSheetBehavior) behaviorField.get(bottomSheetDialog);
            behavior.setBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {

                @Override
                public void onStateChanged(@NonNull View bottomSheet, int newState) {
                    if (newState == BottomSheetBehavior.STATE_DRAGGING) {
                        behavior.setState(BottomSheetBehavior.STATE_EXPANDED);
                    }
                }

                @Override
                public void onSlide(@NonNull View bottomSheet, float slideOffset) {
                }
            });
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Log.d("State", "onCreateDialog");

        BottomSheetDialog dialog = (BottomSheetDialog) super.onCreateDialog(savedInstanceState);
        dialog.setCancelable(false);

        View view = View.inflate(getContext(), R.layout.fragment_lead_details, null);

//        br = new CallStatus();
        firestore = new com.mudrahome.MHLMS.Firebase.Firestore();

        name = view.findViewById(R.id.customer_name);
        loan = view.findViewById(R.id.loan_amount);
        number = view.findViewById(R.id.contact_number);
        employment = view.findViewById(R.id.employment);
        employmentType = view.findViewById(R.id.employment_type);
        loanType = view.findViewById(R.id.loan_type);
        propertyType = view.findViewById(R.id.property_type);
        location = view.findViewById(R.id.location);
        assignedTo = view.findViewById(R.id.assign_to);
        assigner = view.findViewById(R.id.assigner);
        customerRemarks = view.findViewById(R.id.customer_remarks);
        status = view.findViewById(R.id.status);
        workDate = view.findViewById(R.id.date);
        workTime = view.findViewById(R.id.time);
        assignedOn = view.findViewById(R.id.assign_date);
        assignedAt = view.findViewById(R.id.assign_time);
        bankNames = view.findViewById(R.id.bank_names);
        assignerContact = view.findViewById(R.id.assigner_contact_number);
        assigneeContact = view.findViewById(R.id.assignee_contact_number);
        callerRemarksLayout = view.findViewById(R.id.caller_remarks_layout);
        sallerRemarksLayout = view.findViewById(R.id.sales_remarks_layout);
        customerRemarksLayout = view.findViewById(R.id.customer_remarks_layout);

        salesmanRemarksHeadingLayout = view.findViewById(R.id.salesman_remarks_heading_layout);

        employmentTypeLayout = view.findViewById(R.id.employment_type_layout);
        assignedToLayout = view.findViewById(R.id.assigned_to_layout);
        assignerLayout = view.findViewById(R.id.assigner_layout);
        button = view.findViewById(R.id.edit_lead_details);

        setLayoutFields();

        if (leadDetails.getAssignerUId() == null) {
            assignerContact.setClickable(false);
            assignerContact.setCompoundDrawables(null, null, null, null);
            assignerContact.setText("Not available");
        } else {

            try{
                firestore.getUsers(new FirestoreInterfaces.OnGetUserDetails() {
                    @Override
                    public void onSuccess(UserDetails userDetails) {

                            try {
                                assignerContact.setText(userDetails.getContactNumber());
                            }catch (Exception e){
                                assignerContact.setText("Not available");
                            }

                    }

                    @Override
                    public void fail() {

                    }
                },leadDetails.getAssignerUId());
            }catch (Exception e){
                assignerContact.setText("Not available");
            }


            assignerContact.setOnClickListener(view13 -> {

                PermissionManager permission = new PermissionManager(getContext());

                if (permission.checkCallPhone()) {
//                    if (permission.checkReadPhoneState()) {
//                        if (permission.checkRecordAudio()) {
                    callCustomer(leadDetails.getAssignerContact());
//                        } else
//                            permission.requestRecordAudio();
//                    } else
//                        permission.requestReadPhoneState();
                } else
                    permission.requestCallPhone();
            });

        }

        if (leadDetails.getAssignedToUId() == null) {
            /*assigneeCallbutton.setVisibility(View.GONE);*/

            assigneeContact.setClickable(false);
            assigneeContact.setCompoundDrawables(null, null, null, null);
            assigneeContact.setText("Not available");
        } else {

            try {
                firestore.getUsers(new FirestoreInterfaces.OnGetUserDetails() {
                    @Override
                    public void onSuccess(UserDetails userDetails) {

                        try {
                            assigneeContact.setText(userDetails.getContactNumber());
                        }catch (Exception e){
                            assigneeContact.setText("Not available");
                            Log.d("AssigneeContact", " error " + e.getMessage());
                        }



                    }

                    @Override
                    public void fail() {

                    }
                },leadDetails.getAssignedToUId());
            }catch (Exception e){
                assigneeContact.setText("Not available");
            }




            assigneeContact.setOnClickListener(view12 -> {

                PermissionManager permission = new PermissionManager(getContext());

                if (permission.checkCallPhone()) {
//                    if (permission.checkReadPhoneState()) {
//                        if (permission.checkRecordAudio()) {
                    callCustomer(leadDetails.getAssigneeContact());
//                        } else
//                            permission.requestRecordAudio();
//                    } else
//                        permission.requestReadPhoneState();
                } else
                    permission.requestCallPhone();
            });

        }

        number.setOnClickListener(view1 -> {

            PermissionManager permission = new PermissionManager(getContext());

            if (permission.checkCallPhone()) {
//                    if (permission.checkReadPhoneState()) {
//                        if (permission.checkRecordAudio()) {
                callCustomer(leadDetails.getContactNumber());
//                        } else
//                            permission.requestRecordAudio();
//                    } else
//                        permission.requestReadPhoneState();
            } else
                permission.requestCallPhone();
        });

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ConnectivityManager cm = (ConnectivityManager) getActivity()
                        .getSystemService(Context.CONNECTIVITY_SERVICE);

                if (cm.getActiveNetworkInfo() != null) {
                    if (userType.equals(getString(R.string.telecaller))) {
                        progress = new ProgressDialog(context);
                        progress.setMessage("Loading...");
                        progress.setCancelable(false);
                        progress.setCanceledOnTouchOutside(false);
                        progress.show();

                        firestore.fetchUsersByUserType(
                                onFetchSalesPersonList(),
                                leadDetails.getLocation(),
                                getString(R.string.salesman));
                    } else {
                        openSalesmanFragment();
                    }
                } else
                    Toast.makeText(context, R.string.no_internet, Toast.LENGTH_SHORT).show();
            }
        });

        dialog.setContentView(view);
        mBehavior = BottomSheetBehavior.from((View) view.getParent());
        return dialog;
    }

    private void setLayoutFields() {
        if (userType.equals(getString(R.string.telecaller)))
            assignerLayout.setVisibility(View.GONE);
        else if (userType.contains(getString(R.string.salesman)))
            assignedToLayout.setVisibility(View.GONE);
        else
            button.setVisibility(View.GONE);

        if (leadDetails.getEmploymentType().equals("None") || leadDetails.getEmploymentType().isEmpty()) {
            employmentTypeLayout.setVisibility(View.GONE);
        } else {
            employmentType.setText(leadDetails.getEmploymentType());
        }

        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 2);

        sallerRemarksLayout.removeAllViewsInLayout();
        if (leadDetails.getSalesmanReason() == null  || leadDetails.getSalesmanReason().get(0).equals("None")) {
            salesmanRemarksHeadingLayout.setVisibility(View.GONE);
            sallerRemarksLayout.setVisibility(View.GONE);
        } else {

            View view2 = new View(getContext());
            view2.setLayoutParams(layoutParams);
            view2.setBackgroundColor(getResources().getColor(R.color.colorGray));
            sallerRemarksLayout.addView(view2);

            for (int i = leadDetails.getSalesmanReason().size() - 1; i >= 0; i--) {

                if(!leadDetails.getSalesmanReason().get(i).isEmpty()){
                    TextView textView = new TextView(getContext());
                    textView.setTextColor(getResources().getColor(R.color.coloBlack));
                    textView.setText(leadDetails.getSalesmanReason().get(i));
                    sallerRemarksLayout.addView(textView);

                    View view3 = new View(getContext());
                    view3.setLayoutParams(layoutParams);
                    view3.setBackgroundColor(getResources().getColor(R.color.colorGray));
                    sallerRemarksLayout.addView(view3);
                }

            }
        }

        if (leadDetails.getSalesmanRemarks().equals("None") || leadDetails.getSalesmanRemarks().isEmpty()) {
            customerRemarksLayout.setVisibility(View.GONE);
        } else {
            customerRemarks.setText(leadDetails.getSalesmanRemarks());
        }
        name.setText(leadDetails.getName());
        loan.setText(leadDetails.getLoanAmount());
        number.setText(leadDetails.getContactNumber());
        employment.setText(leadDetails.getEmployment());
        loanType.setText(leadDetails.getLoanType());
        propertyType.setText(leadDetails.getPropertyType());
        location.setText(leadDetails.getLocation());
        assignedTo.setText(leadDetails.getAssignedTo());
        assigner.setText(leadDetails.getAssigner());
        status.setText(leadDetails.getStatus());
        workDate.setText(leadDetails.getDate());
        workTime.setText(leadDetails.getTime());
        assignedOn.setText(leadDetails.getAssignDate());
        assignedAt.setText(leadDetails.getAssignTime());



        callerRemarksLayout.removeAllViewsInLayout();



        StringBuilder csvBuilder = new StringBuilder();
        for (String bank : leadDetails.getBanks()) {
            csvBuilder.append(bank);
            csvBuilder.append(", ");
        }
        String bankList = csvBuilder.toString();
        bankNames.setText(bankList);

        View view = new View(getContext());
        view.setLayoutParams(layoutParams);
        view.setBackgroundColor(getResources().getColor(R.color.colorGray));
        callerRemarksLayout.addView(view);

        for (int i = leadDetails.getTelecallerRemarks().size() - 1; i >= 0; i--) {

            if(!leadDetails.getTelecallerRemarks().get(i).isEmpty()){
                TextView textView = new TextView(getContext());
                textView.setTextColor(getResources().getColor(R.color.coloBlack));
                textView.setText(leadDetails.getTelecallerRemarks().get(i));
                callerRemarksLayout.addView(textView);

                View view1 = new View(getContext());
                view1.setLayoutParams(layoutParams);
                view1.setBackgroundColor(getResources().getColor(R.color.colorGray));
                callerRemarksLayout.addView(view1);
            }

        }
    }

    @Override
    public void onStart() {
        super.onStart();
        mBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
//        IntentFilter filter = new IntentFilter("android.intent.action.PHONE_STATE");
//        context.registerReceiver(br, filter);
//        Log.d("State", "onStart");
    }

    private FirestoreInterfaces.OnFetchUsersList onFetchSalesPersonList() {
        return userList -> {
            progress.dismiss();
            if (userList.getUserList().size() > 0)
                openTelecallerFragment(userList.getUserList());
            else
                Toast.makeText(context, "No Salesmen are present for " +
                        leadDetails.getLocation() + ".", Toast.LENGTH_SHORT).show();
        };
    }

    private void openTelecallerFragment(final List<UserDetails> userList) {
        if (userList.size() != 0) {
            TelecallerEditLeadFragment.newInstance(
                    leadDetails, userList, dialogLeadDetails -> {
                        progress = new ProgressDialog(context);
                        progress.setMessage("Loading..");
                        progress.setCancelable(false);
                        progress.setCanceledOnTouchOutside(false);
                        progress.show();

                        firestore.updateLeadDetails(onUpdateLead(), dialogLeadDetails);
                    }).show(getFragmentManager(), "promo");
        }
    }

    private void openSalesmanFragment() {
        SalesmanEditLeadFragment.newInstance(leadDetails, (dialogSalesmanRemarks, dialogSalesmanReason, banks) -> {

            TimeManager timeManager = new TimeManager();
            TimeModel timeModel = timeManager.getTime();

            leadDetails.setDate(timeModel.getDate());
            leadDetails.setTime(timeModel.getTime());
            leadDetails.setTimeStamp(timeModel.getTimeStamp());
            leadDetails.setBanks(banks);

            ArrayList<String> salesmanReson = leadDetails.getSalesmanReason();
            salesmanReson.add(dialogSalesmanReason);

            leadDetails.setSalesmanRemarks(dialogSalesmanRemarks);
            leadDetails.setSalesmanReason(salesmanReson);

            if (dialogSalesmanRemarks.equals(customerNotInterested))
                leadDetails.setStatus(getString(R.string.inactive));
            else if (dialogSalesmanRemarks.equals(documentPicked))
                leadDetails.setStatus(getString(R.string.work_in_progress));
            else if (dialogSalesmanRemarks.equals(documentPickedFileLoggedIn))
                leadDetails.setStatus(getString(R.string.closed));
            else if (dialogSalesmanRemarks.equals(customerFollowUp))
                leadDetails.setStatus(getString(R.string.follow_up));
            else if (dialogSalesmanRemarks.equals(customerNotContactable))
                leadDetails.setStatus(getString(R.string.inactive));
            else if (dialogSalesmanRemarks.equals(customerInterestedButDocumentPending))
                leadDetails.setStatus(getString(R.string.work_in_progress));
            else if (dialogSalesmanRemarks.equals(notDoable))
                leadDetails.setStatus(getString(R.string.not_doable));
            else
                leadDetails.setStatus(getString(R.string.active));

            progress = new ProgressDialog(context);
            progress.setMessage("Loading..");
            progress.setCancelable(false);
            progress.setCanceledOnTouchOutside(false);
            progress.show();

            firestore.updateLeadDetails(onUpdateLead(), leadDetails);
        }).show(getFragmentManager(), "promo");
    }

    private FirestoreInterfaces.OnUpdateLead onUpdateLead() {
        return new FirestoreInterfaces.OnUpdateLead() {
            @Override
            public void onLeadUpdated() {
                Toast.makeText(context, R.string.lead_update, Toast.LENGTH_SHORT).show();
                setLayoutFields();
                if (progress.isShowing())
                    progress.dismiss();
            }

            @Override
            public void onFailer() {
                Toast.makeText(context, R.string.lead_update_failed, Toast.LENGTH_SHORT).show();
            }
        };
    }

    private void callCustomer(String number) {
        Intent callIntent = new Intent(Intent.ACTION_CALL);
        callIntent.setData(Uri.parse("tel:" + number));
        startActivity(callIntent);
    }

//    @Override
//    public void onDestroy() {
//        Log.d("State", "onDestroy");
//        context.unregisterReceiver(br);
//        super.onDestroy();
//    }
}

File: app/src/main/java/com/mudrahome/mhlms/fragments/LeadDetailsFragment.kt
package com.mudrahome.mhlms.fragments

import android.annotation.SuppressLint
import android.app.ProgressDialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.mudrahome.mhlms.R
import com.mudrahome.mhlms.databinding.FragmentLeadDetailsBinding
import com.mudrahome.mhlms.firebase.Firestore
import com.mudrahome.mhlms.fragments.ViewAllRemarksFragment.newInstance
import com.mudrahome.mhlms.interfaces.FirestoreInterfaces
import com.mudrahome.mhlms.interfaces.OnSubmitClickListener
import com.mudrahome.mhlms.managers.PermissionManager
import com.mudrahome.mhlms.managers.ProfileManager
import com.mudrahome.mhlms.managers.TimeManager
import com.mudrahome.mhlms.model.LeadDetails
import com.mudrahome.mhlms.model.UserDetails
import com.mudrahome.mhlms.model.UserList
import java.util.*

@SuppressLint("ValidFragment")
class LeadDetailsFragment(
    private val leadDetails: LeadDetails,
    private val userType: String
) : BottomSheetDialogFragment() {

    private var progress: ProgressDialog? = null
    private var firestore: Firestore? = null
    private var manager: ProfileManager? = null

    private val customerNotInterested = "Customer Not Interested"
    private val documentPicked = "Document Picked"
    private val customerFollowUp = "Customer follow Up"
    private val customerNotContactable = "Customer Not Contactable"
    private val customerInterestedButDocumentPending = "Customer Interested but Document Pending"
    private val notDoable = "Not Doable"
    private val documentPickedFileLoggedIn = "Document Picked and File Logged in"

    private var isEdit: Boolean? = false
    private var binding: FragmentLeadDetailsBinding? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(
            LayoutInflater.from(context),
            R.layout.fragment_lead_details, null, false
        )

        manager = ProfileManager(context)
        firestore = Firestore()

        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (userType.equals(context!!.getString(R.string.admin))) {
            binding!!.editLeadButton.visibility = View.GONE
        }

        binding!!.editLeadButton.setOnClickListener {
            setButtonAction()
        }

        setText()

        if (leadDetails.assignerUId != null) {
            firestore!!.getUsers(object : FirestoreInterfaces.OnGetUserDetails {
                override fun onSuccess(userDetails: UserDetails?) {

                    if (userDetails?.contactNumber == null) {
                        hideCallerContact()
                    } else {
                        binding!!.callerContactNumber!!.text = userDetails.contactNumber
                        binding!!.callerContactNumber.isClickable = true
                    }
                }

                override fun fail() {
                    hideCallerContact()
                }

            }, leadDetails.assignerUId!!)
        }

        if (leadDetails.assignedToUId != null) {
            firestore!!.getUsers(object : FirestoreInterfaces.OnGetUserDetails {
                override fun onSuccess(userDetails: UserDetails?) {

                    if (userDetails?.contactNumber == null) {
                        hideAssigneeContact()
                    } else {
                        binding!!.assigneeContactNumber!!.text = userDetails.contactNumber
                        binding!!.assigneeContactNumber.isClickable = true
                    }
                }

                override fun fail() {
                    hideAssigneeContact()
                }

            }, leadDetails.assignedToUId!!)
        }

        if (leadDetails.forwarderUId != null) {
            firestore!!.getUsers(object : FirestoreInterfaces.OnGetUserDetails {
                override fun onSuccess(userDetails: UserDetails?) {

                    if (userDetails?.contactNumber == null) {
                        hideAssignerContact()
                    } else {
                        binding!!.assignerContactNumber!!.text = userDetails.contactNumber
                        binding!!.assignerContactNumber.isClickable = true

                    }
                }

                override fun fail() {
                    hideAssignerContact()
                }

            }, leadDetails.forwarderUId!!)
        }

        binding!!.callerContactNumber.setOnClickListener {
            val permission = PermissionManager(context)
            if (permission.checkCallPhone()) {
                callCustomer(binding!!.callerContactNumber.text.toString())
            } else
                permission.requestCallPhone()
        }

        binding!!.assignerContactNumber.setOnClickListener {
            val permission = PermissionManager(context)
            if (permission.checkCallPhone()) {
                callCustomer(binding!!.assignerContactNumber.text.toString())
            } else
                permission.requestCallPhone()
        }

        binding!!.assigneeContactNumber.setOnClickListener {
            val permission = PermissionManager(context)
            if (permission.checkCallPhone()) {
                callCustomer(binding!!.assigneeContactNumber.text.toString())
            } else
                permission.requestCallPhone()
        }
    }

    private fun hideAssignerContact() {
        binding!!.assignerContactNumber.isClickable = false
        binding!!.assignerContactNumber.setCompoundDrawables(null, null, null, null)
        binding!!.assignerContactNumber.text = "Not available"
    }

    private fun hideAssigneeContact() {
        binding!!.assigneeContactNumber.isClickable = false
        binding!!.assigneeContactNumber.setCompoundDrawables(null, null, null, null)
        binding!!.assigneeContactNumber.text = "Not available"
    }

    private fun hideCallerContact() {
        binding!!.callerContactNumber.isClickable = false
        binding!!.callerContactNumber.setCompoundDrawables(null, null, null, null)
        binding!!.callerContactNumber.text = "Not available"
    }

    private fun setText() {
        binding!!.customerName.text = leadDetails.name
        binding!!.loanAmount.text = leadDetails.loanAmount
        binding!!.contactNumber.text = leadDetails.contactNumber
        binding!!.employment.text = leadDetails.employment
        binding!!.employmentType.text = leadDetails.employmentType
        binding!!.loanType.text = leadDetails.loanType
        binding!!.propertyType.text = leadDetails.propertyType
        binding!!.location.text = leadDetails.location
        binding!!.caller!!.text = leadDetails.assigner
        binding!!.assigner!!.text = leadDetails.forwarderName
        binding!!.customerRemarks.text = leadDetails.salesmanRemarks
        binding!!.date.text = Date(leadDetails.timeStamp).toString()
        binding!!.status.text = leadDetails.status


        if (leadDetails.contactNumber == null) {
            binding!!.contactNumber.isClickable = false
            binding!!.contactNumber.setCompoundDrawables(null, null, null, null)
        } else {
            binding!!.contactNumber.setOnClickListener {
                val permission = PermissionManager(context)
                if (permission.checkCallPhone()) {
                    callCustomer(binding!!.contactNumber.text.toString())
                } else
                    permission.requestCallPhone()
            }
        }

        if (leadDetails.assignedTo.isNullOrEmpty()) {
            binding!!.assignTo.text = "Not assigned yet"
            binding!!.assignDate.text = "Not assigned yet"
            binding!!.assignTime.text = "Not assigned yet"

            hideAssigneeContact()
        } else {
            binding!!.assignTo.text = leadDetails.assignedTo
            binding!!.status.text = leadDetails.status
            binding!!.assignDate.text = leadDetails.assignDate
            binding!!.assignTime.text = leadDetails.assignTime

        }

        if (leadDetails.banks.isNullOrEmpty()) {
            binding!!.bankLayout.visibility = View.GONE
        } else {
            val csvBuilder = StringBuilder()
            for (bank in leadDetails.banks) {
                csvBuilder.append(bank)
                csvBuilder.append(", ")
            }
            val bankList = csvBuilder.toString()
            binding!!.bankNames.text = bankList
            binding!!.bankLayout.visibility = View.VISIBLE
        }

        if (!leadDetails.salesmanReason.isNullOrEmpty()) {
            binding!!.assigneeRemarks.setOnClickListener { view ->
                newInstance(leadDetails.salesmanReason, "Salesman's").show(
                    childFragmentManager,
                    "ShowRemarks"
                )
            }
        }

        if (!leadDetails.telecallerRemarks.isNullOrEmpty()) {
            binding!!.callerRemarks!!.setOnClickListener { view ->
                newInstance(leadDetails.telecallerRemarks, "Caller's").show(
                    childFragmentManager,
                    "ShowRemarks"
                )
            }
        }

        if (!leadDetails.forwarderRemarks.isNullOrEmpty()) {
            binding!!.assignerRemarks!!.setOnClickListener { view ->
                newInstance(leadDetails.forwarderRemarks, "Caller's").show(
                    childFragmentManager,
                    "ShowRemarks"
                )
            }
        }
    }

    private fun setButtonAction() {
        when {
            leadDetails.status == "Incomplete Lead" -> {
                val transaction =
                    (context as AppCompatActivity).supportFragmentManager.beginTransaction()
                val fragment: Fragment = UploadLeadFragment()

                val bundle = Bundle()
                bundle.putSerializable("leadDetails", leadDetails)
                fragment.arguments = bundle

                transaction.addToBackStack(null)
                transaction.replace(R.id.content_frame, fragment)
                transaction.addToBackStack(null)
                transaction.commit()
            }

            manager!!.spUserType == context!!.getString(R.string.telecaller) && leadDetails.forwarderUId != null -> {
                openCallerFragment()
            }

            manager!!.spUserType == context!!.getString(R.string.telecaller_and_teleassigner) ||
                    manager!!.spUserType == context!!.getString(R.string.telecaller) -> {
                firestore!!.fetchUsersByUserType(
                    onFetchSalesPersonList(),
                    leadDetails.location!!,
                    context!!.getString(R.string.salesman)
                )
            }

            userType == context!!.getString(R.string.salesman) -> {
                openSalesmanFragment()
            }

            else -> {
                binding!!.editLeadButton.visibility = View.GONE
            }
        }
    }

    private fun onFetchSalesPersonList(): FirestoreInterfaces.OnFetchUsersList {
        return object : FirestoreInterfaces.OnFetchUsersList {
            override fun onFail() {
                progress?.dismiss()
            }

            override fun onListFetched(userList: UserList?) {
                progress?.dismiss()

                if (userList!!.userList.size > 0) {
                    openAssignerFragment(userList.userList)
                } else {
                    Toast.makeText(
                        context,
                        "No salesman present for this location.",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }

        }
    }

    private fun openCallerFragment() {
        CallerEditLeadFragment(
            leadDetails, OnSubmitClickListener { dialogLeadDetails ->
                progress = ProgressDialog(context)
                progress!!.setMessage("Loading..")
                progress!!.setCancelable(false)
                progress!!.setCanceledOnTouchOutside(false)
                progress!!.show()

                firestore!!.updateLeadDetails(onUpdateLead(), dialogLeadDetails)
            }
        ).show(fragmentManager!!, "promo")
    }

    private fun openAssignerFragment(userList: List<UserDetails>) {
        AssignerEditLeadFragment(
            leadDetails, userList, userType, OnSubmitClickListener { dialogLeadDetails ->
                progress = ProgressDialog(context)
                progress!!.setMessage("Loading..")
                progress!!.setCancelable(false)
                progress!!.setCanceledOnTouchOutside(false)
                progress!!.show()

                leadDetails.status = setStatus(leadDetails.salesmanRemarks!!)

                firestore!!.updateLeadDetails(onUpdateLead(), dialogLeadDetails)
            }
        ).show(fragmentManager!!, "promo")
    }

    private fun openSalesmanFragment() {
        SalesmanEditLeadFragment.newInstance(leadDetails) { dialogSalesmanRemarks, dialogSalesmanReason, banks ->

            val timeManager = TimeManager()

            //            leadDetails.setDate(timeModel.getDate());
            //            leadDetails.setTime(timeModel.getTime());
            leadDetails.timeStamp = timeManager.timeStamp
            leadDetails.banks = banks

            val salesmanReson = leadDetails.salesmanReason
            salesmanReson.add(dialogSalesmanReason + "@@" + System.currentTimeMillis())        // Set SalesmanReason with timesteamp

            leadDetails.salesmanRemarks = dialogSalesmanRemarks
            leadDetails.salesmanReason = salesmanReson
            leadDetails.status = setStatus(dialogSalesmanRemarks)

            progress = ProgressDialog(context)
            progress!!.setMessage("Loading..")
            progress!!.setCancelable(false)
            progress!!.setCanceledOnTouchOutside(false)
            progress!!.show()
            firestore!!.updateLeadDetails(onUpdateLead(), leadDetails)
        }.show(fragmentManager!!, "promo")
    }

    private fun setStatus(remarks: String): String {
        if (remarks == customerNotInterested)
            return getString(R.string.inactive)
        else if (remarks == documentPicked)
            return getString(R.string.work_in_progress)
        else if (remarks == documentPickedFileLoggedIn)
            return getString(R.string.closed)
        else if (remarks == customerFollowUp)
            return getString(R.string.follow_up)
        else if (remarks == customerNotContactable)
            return getString(R.string.inactive)
        else if (remarks == customerInterestedButDocumentPending)
            return getString(R.string.work_in_progress)
        else if (remarks == notDoable)
            return getString(R.string.not_doable)
        else
            return getString(R.string.active)
    }

    private fun onUpdateLead(): FirestoreInterfaces.OnUpdateLead {
        return object : FirestoreInterfaces.OnUpdateLead {
            override fun onLeadUpdated() {
                Toast.makeText(context, R.string.lead_update, Toast.LENGTH_SHORT).show()
                isEdit = true
                setText()
                if (progress!!.isShowing)
                    progress!!.dismiss()
            }

            override fun onFailer() {
                Toast.makeText(context, R.string.lead_update_failed, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun callCustomer(number: String?) {
        val callIntent = Intent(Intent.ACTION_CALL)
        callIntent.data = Uri.parse("tel:" + number!!)
        startActivity(callIntent)
    }
}